### ISSUE TITLE
Compile error when switching between branches from new format back to branch with old format

### ISSUE THREAD

**Issue by chrisaut** [Microsoft/msbuild#3097]
I'm trying to upgrade some libraries in a big solution, it's a step by step process, one lib at a time, so I'm using a separate branch along the way. When switching back to main branch to do some unrelated work I noticed nothing would compile anymore.

Note I'm not sure if msbuild is the correct place to file this issue, the error comes from the nuget.target.

### Steps to reproduce
 - Create a new VS project (Class Library, .NET Framework) 4.5.2
 - Make sure it compiles, commit changes to git
 - create new branch "convert" (or whatever)
 - delete the assemblyinfo.cs
 - Unload project, replace all with
```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
  </PropertyGroup>
</Project>
```
- Reload, make sure it compiles
- Commit to this branch
- switch branches back to master (or whatever the branch was)
- Try to compile
- Get angry because it says 
```1>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Microsoft\NuGet\15.0\Microsoft.NuGet.targets(186,5): error : Your project is not referencing the ".NETFramework,Version=v4.5.2" framework. Add a reference to ".NETFramework,Version=v4.5.2" in the "frameworks" section of your project.json, and then re-run NuGet restore.```
 - I thought VS cached something, but even deleting the .vs folder completely and reoping the solution still leaves me unable to compile my project.
 - Note that switching to the netstandard branch again compiles fine

OS info:

Win 10, VS 15.6.2 Professional

**Comment by jp2masa** [comment]
I think that deleting the "obj" folder should solve that, as it forces a new restore. I think that the problem is that project.assets.json is treated as up-to-date, although it's not, but not sure.

**Comment by chrisaut** [comment]
Yes, deleting the obj folder fixes this. Note specifically its the "project.assets.json" file inside there, this gets created as soon as I compile the netstandard version, switching back to regular .NET with this file present triggers this issue.

Perhaps the simplest solution would be to ensure that a "Clean" operation actually deletes this file.

**Comment by jp2masa** [comment]
Not sure, but I think that "project.assets.json" is used by UWP projects too, so it can't be always removed, but maybe adding a check for "project.json" and removing "project.assets.json" if it doesn't exist would work.

**Comment by chrisaut** [comment]
Sorry to bump, but is this something the team is considering fixing or should we just live with it during the upgrading of projects? Should I report this somewhere else?

**Comment by rainersigwald** [comment]
I usually just do a `git clean -xdf` after switching past that boundary, which is definitely suboptimal, but I don't think there's an easy way out of this.

@nguerrera @rohit21agrawal can you think of anything we can do to make this smoother?

**Comment by rohit21agrawal** [comment]
@chrisaut  doing a rebuild should also trigger a new restore as far as I remember

**Comment by chrisaut** [comment]
@rohit21agrawal Unfortunatly not, doing a rebuild does not fix it. That's pretty much the first thing I tried. The user has to figure out that the project.assets.json is to blame and manually delete the file every time they switch branches.

**Comment by duncansmart** [comment]
Been using this in `.git/hooks/post-checkout` which automatically deletes project.assets.json when you switch branches:

	#!/bin/sh
	# Args:
	#   $1: Ref of previous HEAD
	#   $2: Ref of new HEAD
	#   $3: Whether this is a file checkout (0) or branch checkout (1)

	# if this is a file checkout – do nothing
	if [ "$3" == "0" ]; then exit; fi

	find -maxdepth 3 -type f -path "*/obj/project.assets.json" -delete

You might need to tweak `-maxdepth` according to how deeply you nest projects

**Comment by yorkadam** [comment]
We are having the same issue.  However, it's not just the "obj" folder.  It is also other items like projects in the solution that are left behind.  We are moving some legacy applications from VB.NET to C# and .NET Core and switching between branches causes the error in reported above but also leaves the local folder with mixed C# and VB folders.  We are using Azure Repos and Visual Studio 2019.

Example:
1. Create a new branch from existing VB solution that contains more than one projects.
2. Check out the new branch
3. Replace with C# files and remove outdated projects in solution.
4. Check in.
5. Checkout old VB branch - Error same as above regarding "obj" folder.
6.  Switch back (check out) to new branch (1. above).
7. Local folder now contains extra projects that were not removed during the check-out process.
   In this case, the left over projects are from the VB branch that should have been removed when checking out C# branch.

Our expectation on "check-out" between different branches that all files would be removed before the new branch is loaded.
Our work around for now is to create a new repository instead of new branches.

Thanks
